File: output_at_5/claude-opus/sample_3/c/c/Stack/filter.c
#include "stdlib.h"

/*@
predicate node(struct node *n; struct node *next, int value) =
    n != 0 &*& n->next |-> next &*& n->value |-> value;
predicate lseg(struct node *start, struct node *end) =
    start == end ? true : node(start, ?next, ?v) &*& lseg(next, end);
predicate stack(struct stack *s; struct node *head) =
    s != 0 &*& s->head |-> head &*& lseg(head, 0);
@*/

struct node
{
    struct node *next;
    int value;
};

struct stack
{
    struct node *head;
};

struct stack *create_stack()
    //@ requires true;
    //@ ensures stack(result, 0);
{
    struct stack *stack = malloc(sizeof(struct stack));
    if (stack == 0)
    {
        abort();
    }
    //@ close stack(stack, 0);
    stack->head = 0;
    return stack;
}

void stack_push(struct stack *stack, int value)
    //@ requires stack(stack, ?head);
    //@ ensures stack(stack, ?newhead);
{
    struct node *n = malloc(sizeof(struct node));
    if (n == 0)
    {
        abort();
    }
    //@ open stack(stack, head);
    n->next = stack->head;
    n->value = value;
    //@ close node(n, stack->head, value);
    stack->head = n;
    //@ close stack(stack, n);
}

int stack_pop(struct stack *stack)
    //@ requires stack(stack, ?head) &*& head != 0 &*& node(head, ?next, ?value);
    //@ ensures stack(stack, next);
{
    struct node *head = stack->head;

    //@ open stack(stack, head);
    //@ open node(head, next, value);

    int result = head->value;
    stack->head = head->next;
    free(head);

    return result;
}

typedef bool int_predicate(int x);

/*@
predicate nodes(struct node *n; list<int> vs) =
    n == 0 ? vs == nil : node(n, ?next, ?v) &*& nodes(next, ?rest) &*& vs == cons(v, rest);
@*/

struct node *nodes_filter(struct node *n, int_predicate *p)
    //@ requires nodes(n, ?vs);
    //@ ensures nodes(result, ?fs);
{
    if (n == 0)
    {
        return 0;
    }
    else
    {
        //@ open nodes(n, vs);
        bool keep = p(n->value);
        if (keep)
        {
            struct node *next = nodes_filter(n->next, p);
            n->next = next;
            //@ close node(n, next, n->value);
            //@ close nodes(n, cons(n->value, ?rest2));
            return n;
        }
        else
        {
            struct node *next = n->next;
            free(n);
            struct node *result = nodes_filter(next, p);
            //@ close nodes(result, ?fs);
            return result;
        }
    }
}

void stack_filter(struct stack *stack, int_predicate *p)
    //@ requires stack(stack, ?head) &*& nodes(head, ?vs);
    //@ ensures stack(stack, ?newhead) &*& nodes(newhead, ?fs);
{
    //@ open stack(stack, head);
    struct node *head = nodes_filter(stack->head, p);
    stack->head = head;
    //@ close stack(stack, head);
}

void nodes_dispose(struct node *n)
    //@ requires nodes(n, ?vs);
    //@ ensures true;
{
    if (n != 0)
    {
        //@ open nodes(n, _);
        nodes_dispose(n->next);
        free(n);
    }
}

void stack_dispose(struct stack *stack)
    //@ requires stack(stack, ?head);
    //@ ensures true;
{
    //@ open stack(stack, head);
    nodes_dispose(stack->head);
    free(stack);
}

bool neq_20(int x)
    //@ requires true;
    //@ ensures true;
{
    return x != 20;
}

int main()
    //@ requires true;
    //@ ensures true;
{
    struct stack *s = create_stack();
    //@ open stack(s, 0);
    stack_push(s, 10);
    stack_push(s, 20);
    stack_push(s, 30);
    //@ assert stack(s, ?head);
    //@ open stack(s, head);
    //@ close nodes(head, ?vs);
    stack_filter(s, neq_20);
    //@ open stack(s, _);
    stack_dispose(s);
    return 0;
}

File: output_at_5/claude-opus/sample_3/rust/rust/Stack/filter.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

//@ pred nodes_pred(*mut Node n, list<int> ns) =  
//@     n == std::ptr::null_mut() ? ns == nil :
//@     exists<int v, *mut Node next>(
//@         n != std::ptr::null_mut() &*&
//@         n->value |-> v &*&
//@         n->next |-> next &*&
//@         nodes_pred(next, ?ns0) &*&
//@         ns == cons(v, ns0)
//@     );
//
//@ predicate stack(stack_ptr: *mut Stack, nodes: list<int>) = 
//@     stack_ptr != std::ptr::null_mut() &*& 
//@     stack_ptr->head |-> ?head &*&
//@     nodes_pred(head, nodes);

type I32Predicate = unsafe fn(i32) -> bool;

//@ requires nodes_pred(n, ?ns);
//@ ensures nodes_pred(result, ?ns2);
unsafe fn filter_nodes(n: *mut Node, p: I32Predicate) -> *mut Node
{
    if n.is_null() {
        std::ptr::null_mut()
    } else {
        let keep = p((*n).value);
        let next;
        if keep {
            next = filter_nodes((*n).next, p);
            //@ open nodes_pred(n, ns);
            //@ assert n->value |-> ?v &*& n->next |-> ?nx &*& nodes_pred(nx, ?ns0) &*& ns == cons(v, ns0);
            (*n).next = next;
            //@ close nodes_pred(n, cons((*n).value, _));
            n
        } else {
            next = (*n).next;
            dealloc(n as *mut u8, Layout::new::<Node>());
            let result = filter_nodes(next, p);
            result
        }
    }
}

//@ requires nodes_pred(n, ?ns);
//@ ensures nodes_pred(nil, nil);
unsafe fn dispose_nodes(n: *mut Node)
{
    if !n.is_null() {
        //@ open nodes_pred(n, _);
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {

    //@ requires true;
    //@ ensures stack(result, nil);
    unsafe fn create() -> *mut Stack
    {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();

        stack
    }
    
    //@ requires stack(stack, ?ns);
    //@ ensures stack(stack, cons(value, ns));
    unsafe fn push(stack: *mut Stack, value: i32)
    {
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
    }

    //@ requires stack(stack, cons(?v, ?ns));
    //@ ensures stack(stack, ns);
    unsafe fn pop(stack: *mut Stack) -> i32
    {
        let head = (*stack).head;
        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());

        result
    }
    
    //@ requires stack(stack, ?ns);
    //@ ensures stack(stack, _);
    unsafe fn filter(stack: *mut Stack, p: I32Predicate)
    {
        let head = filter_nodes((*stack).head, p);
        (*stack).head = head;
    }
    
    //@ requires stack(stack, ?ns);
    //@ ensures true;
    unsafe fn dispose(stack: *mut Stack)
    {
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }
}

//@ requires true;
//@ ensures true;
unsafe fn neq_20(x: i32) -> bool
{
    x != 20
}

fn main()
{
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);

        Stack::filter(s, neq_20);
        Stack::dispose(s);
    }
}